Re-Initiate transactions export in server via download button (un-revert)

chickenn00dle opened the pull request:
This undos the revert of [#1178](https://github.com/Automattic/woocommerce-payments/pull/2939)

#2939 was reverted since the server changes it depended on were causing issues with PHPUnit tests which were only discovered during the deploy process. Since it was the end of the week and I didn't have time to look further into this, I simply reverted so as not to cause problems for others working on Server.

This re-introduces the client changes from #2939 

### Testing Instructions

- If possible sync a connected account with `< 100` transactions
- Via WP-Admin go to Payments > Transactions
- Expand the number of rows in the list to 100, then [click the download link](https://cloudup.com/ccse5g7T_wc)
- Ensure a CSV containing the transactions listed in the table view immediately downloads
- If possible sync a connected account with `> 10000` transactions. If this is not possible, you may need to change the value of [`confirmThreshold`](https://github.com/Automattic/woocommerce-payments/blob/c84f2524b7a309adc80e6d9db531bde35117f09e/client/transactions/list/index.tsx#L418) to be less than the total number of transactions in the connected site while testing.
- Again, click the download link, making sure the number of rows is `<` the total number of transactions to be exported.
- Ensure a confirmation window appears asking if you would like to proceed. Confirm.
- Ensure a notice appears letting you know an export will be emailed to you
- Confirm you receive an email containing a download link for an export containing transactions that meet the criteria
- Select any filter (i.e. before December 1) and click the download link
- Ensure no confirmation window appears
- Ensure a notice appears letting you know an export will be emailed to you
- Confirm you receive an email containing a download link for an export containing transactions that meet the criteria
- Repeat the steps above with different combinations of filters

chickenn00dle commented:
Hey again @naman03malhotra 

I ran into issues while trying to deploy the server PR for the transactions export changes, so had to put this on hold a bit. Now that that has been resolved, this is almost ready to go up again. 

However, @dechov pointed out some optimizations (p1638907040039900/1638560413.495400-slack-CGGCLBN58) we could implement to make the export process less confusing for merchants, and rather than make these changes in a follow-up PR, I thought I'd just get them in here now.

Would you or Paul mind having another look at this. I've re-added and updated the testing instructions in this PR description above.

chickenn00dle commented on client/transactions/list/index.tsx:
```diff
@@ -383,20 +388,100 @@ export const TransactionsList = (
+				!! typeIs ||
+				!! typeIsNot;
+
+			const confirmThreshold = 1000;
```
I arbitrarily set this to be 1000, but am happy to bump this up to something higher. I'm not really sure what the typical WCPay merchant might have here.

chickenn00dle replied:
> why do we need to set a threshold on the client-side

Good question! The threshold is to give merchants a chance to filter prior to kicking off the export. For example, a merchant might click download believing they don't have so many transactions. If their export size is larger than the threshold the merchant can rethink this decision and opt to add some filters for a potentially smaller export.

>  IMO the server should be able to handle, 1k - 10k export easily, wdyt?

Yeah. That was my initial hunch. 1K was a bit too conservative. I've upped this to 10k in https://github.com/Automattic/woocommerce-payments/pull/3477/commits/aa7647f2a055b23da9341ba3413e8d4b059ae13f

chickenn00dle replied:
Sorry, I realized I forgot to address this question directly: 

> what if the client has more than 1000 transactions?

If the merchant has over `confirmThreshold` transactions, they can either confirm and get all of these in their export, OR cancel and select some filters to reduce the size of their export.

chickenn00dle commented:
Just noting that the server changes for this have been merged to `trunk` so there is no longer a need to check out anything else to test this.

htdat commented on client/transactions/list/test/index.tsx:
```diff
@@ -415,16 +426,28 @@ describe( 'Transactions list', () => {
-		afterEach( () => {
-			jest.resetAllMocks();
-		} );
+		test( 'should request confirmation when the download button is clicked', () => {
```
May use a better name here to indicate that the confirmation popup will only display if the to-be-downloaded number of transactions is bigger than the threshold 1000?

chickenn00dle replied:
I changed the names of this test (and the new test to cover the case when merchants choose not to proceed): https://github.com/Automattic/woocommerce-payments/pull/3477/commits/aa7647f2a055b23da9341ba3413e8d4b059ae13f

htdat commented on client/transactions/list/test/index.tsx:
```diff
@@ -415,16 +426,28 @@ describe( 'Transactions list', () => {
+			getByRole( 'button', { name: 'Download' } ).click();
+
+			expect( window.confirm ).toHaveBeenCalledTimes( 1 );
+			expect( window.confirm ).toHaveBeenCalledWith(
```
May we mock the windown.confirm value, and verify that the code will trigger this API request too? 

https://github.com/Automattic/woocommerce-payments/blob/bebc1d686c8287e5408f671d28e687341b0e6ab7/client/transactions/list/index.tsx#L435-L436

naman03malhotra replied:
IMO, if we can mock `apiFetch` and verify the parameters with which it has been called, that would be great.

chickenn00dle replied:
Sure thing! I made these changes in https://github.com/Automattic/woocommerce-payments/pull/3477/commits/aa7647f2a055b23da9341ba3413e8d4b059ae13f

htdat reviewed (commented):
I followed up the test instruction and they worked well. 
The change looks great to me. I've just left some comments to improve the test a bit.

naman03malhotra commented on client/transactions/list/index.tsx:
```diff
@@ -383,20 +388,100 @@ export const TransactionsList = (
-			generateCSVFileName( title, params ),
-			generateCSVDataFromTable( columnsToDisplay, rows )
-		);
+			if (
+				isFiltered ||
+				totalRows < confirmThreshold ||
+				window.confirm( confirmMessage )
```
Can we add a unit test that covers the else case? Behavioral test for the case when these conditions are not met.

chickenn00dle replied:
Absolutely! Done in https://github.com/Automattic/woocommerce-payments/pull/3477/commits/aa7647f2a055b23da9341ba3413e8d4b059ae13f

naman03malhotra reviewed (commented):
Tests well, thanks for working on this. The changes look good, though I've left some questions regarding the threshold and tests.

chickenn00dle commented on changelog.txt:
```diff
@@ -5,6 +5,7 @@
 * Add - JS error boundaries to admin screens.
 * Update - Remove task from the overview list for setting up multiple currencies.
 * Update - Return to task "Set up payments" after finishing KYC from WC-Admin.
+* Add - Add support for full transaction exports.
```
Good catch 👍 

Added in https://github.com/Automattic/woocommerce-payments/pull/3477/commits/1a661a8c1047cdcc72cc0e3959cb89057ba209e8

chickenn00dle commented:
Thanks for the review @htdat and @naman03malhotra!

I've addressed your feedback. This is good for another round of review 🙇

htdat reviewed (approved):
LGTM! But I suggest waiting for @naman03malhotra's review too.

A few things to consider:

1. ~If no rush, let's target this PR for 3.6.0 (scheduled for Jan 19, 2022) PaJDYF-1gY-p2. The main reason is that we have more dev coverage for minor releases if something goes wrong.~
Update: Just found out this comment paJDYF-2WB-p2#comment-9101 so please ignore this consideration.

2. We may test all the test cases but let's add as many as possible for GlobalStep https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions

htdat commented:
> I'll await a reply from @naman03malhotra before merging this one.

Naman will be back next week (Jan 10). If we get it merged next week, we will have only one test round from GlobalStep. 

Thus, I suggest merging the PR and preparing tests for GlobalStep within this week, then we will have two test rounds. 

If Naman has more comments, you can address it later. WDYT about this approach, @chickenn00dle ?

chickenn00dle commented:
> I suggest merging the PR and preparing tests for GlobalStep within this week, then we will have two test rounds.

Given Naman is AFK, this approach works for me. I'll go ahead and merge this now, then get some testing instructions added for 3.6.0

chickenn00dle commented:
Just noting I've added testing instructions here. I'm not sure what GS is able to do in terms of syncing connected accounts, so I tried to split up the testing instructions into sections that they can do or skip accordingly: https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-3.6.0

htdat commented:
> Just noting I've added testing instructions here. 

Can you add a note that they should run these tests as the last ones? With that, they will usually have a few transactions already after running these critical flows https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows. 

>  I'm not sure what GS is able to do in terms of syncing connected accounts.

GlobalStep can not do this on their own. Even for us, it's not easy to do with the production server. Therefore, I think we should not mention it. Otherwise, that only makes them confused.

dechov commented:
If we have an account we want them to test with, maybe we can just make a test site available to them that’s already linked?

chickenn00dle commented:
> Can you add a note that they should run these tests as the last ones?

Sure thing @htdat! I've added the note to [the testing instructions](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-3.6.0).

> GlobalStep can not do this on their own. Even for us, it's not easy to do with the production server.

Gotcha. I've removed the testing instructions for endpoint exports then.

> If we have an account we want them to test with, maybe we can just make a test site available to them that’s already linked?

This sounds good to me. I'm happy to set this up on my end. What is the process for this with GS? Is a Jurassic Ninja site good for this? And if so, am I okay to just search connected accounts for one that has a larger number of transactions?

htdat commented:
> What is the process for this with GS? Is a Jurassic Ninja site good for this? 

Yes, GS uses Jurassic Ninja sites for testing.

> And if so, am I okay to just search connected accounts for one that has a larger number of transactions?

As long as it's a test account, I think that should be no issue.

chickenn00dle commented:
> Yes, GS uses Jurassic Ninja sites for testing.

Thanks for clarifying. I've set up a test site for this and added test instructions for the endpoint export (< 10k transactions). I'm going to ping you directly @htdat to find out what the best practice is for sharing passwords for test sites such as these.

> As long as it's a test account, I think that should be no issue.

Great. I had a look through the list of test accounts and ended up selecting `acct_1FgwJeIdEKZt6HdP` since it had the largest number of transactions and uses one of our email addresses. This account is associated with @marcinbot. So this means the export email will go to your A8C email address when GS runs [this test](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-3.6.0#endpoint-exports--10000-transactions). Would it be possible to confirm this here afterward?

dechov commented:
I think the old E2E testing account `acct_1H5ZCdK66Zjn17SG` might have over 10k transactions but I'm not sure.

htdat commented:
> This account is associated with @marcinbot so I think this means the export email will go to your A8C email address when GS runs this test. Would it be possible to confirm this here afterward?

I think it will go to the WPcom email address that owns the Jetpack connection of the test site, aka it's your email @chickenn00dle for this blog_id 201497760.
